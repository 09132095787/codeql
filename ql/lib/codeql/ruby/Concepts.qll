/**
 * Provides abstract classes representing generic concepts such as file system
 * access or system command execution, for which individual framework libraries
 * provide concrete subclasses.
 */

private import codeql.ruby.AST
private import codeql.ruby.CFG
private import codeql.ruby.DataFlow
private import codeql.ruby.Frameworks
private import codeql.ruby.dataflow.RemoteFlowSources
private import codeql.ruby.ApiGraphs

/**
 * A data-flow node that executes SQL statements.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `SqlExecution::Range` instead.
 */
class SqlExecution extends DataFlow::Node instanceof SqlExecution::Range {
  /** Gets the argument that specifies the SQL statements to be executed. */
  DataFlow::Node getSql() { result = super.getSql() }
}

/** Provides a class for modeling new SQL execution APIs. */
module SqlExecution {
  /**
   * A data-flow node that executes SQL statements.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `SqlExecution` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Gets the argument that specifies the SQL statements to be executed. */
    abstract DataFlow::Node getSql();
  }
}

/**
 * A data-flow node that escapes meta-characters, which could be used to prevent
 * injection attacks.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `Escaping::Range` instead.
 */
class Escaping extends DataFlow::Node instanceof Escaping::Range {
  Escaping() {
    // escapes that don't have _both_ input/output defined are not valid
    exists(super.getAnInput()) and
    exists(super.getOutput())
  }

  /** Gets an input that will be escaped. */
  DataFlow::Node getAnInput() { result = super.getAnInput() }

  /** Gets the output that contains the escaped data. */
  DataFlow::Node getOutput() { result = super.getOutput() }

  /**
   * Gets the context that this function escapes for, such as `html`, or `url`.
   */
  string getKind() { result = super.getKind() }
}

/** Provides a class for modeling new escaping APIs. */
module Escaping {
  /**
   * A data-flow node that escapes meta-characters, which could be used to prevent
   * injection attacks.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `Escaping` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Gets an input that will be escaped. */
    abstract DataFlow::Node getAnInput();

    /** Gets the output that contains the escaped data. */
    abstract DataFlow::Node getOutput();

    /**
     * Gets the context that this function escapes for.
     *
     * While kinds are represented as strings, this should not be relied upon. Use the
     * predicates in  the `Escaping` module, such as `getHtmlKind`.
     */
    abstract string getKind();
  }

  /** Gets the escape-kind for escaping a string so it can safely be included in HTML. */
  string getHtmlKind() { result = "html" }
}

/**
 * An escape of a string so it can be safely included in
 * the body of an HTML element, for example, replacing `{}` in
 * `<p>{}</p>`.
 */
class HtmlEscaping extends Escaping {
  HtmlEscaping() { super.getKind() = Escaping::getHtmlKind() }
}

/** Provides classes for modeling HTTP-related APIs. */
module HTTP {
  /** Provides classes for modeling HTTP servers. */
  module Server {
    /**
     * A data-flow node that sets up a route on a server.
     *
     * Extend this class to refine existing API models. If you want to model new APIs,
     * extend `RouteSetup::Range` instead.
     */
    class RouteSetup extends DataFlow::Node instanceof RouteSetup::Range {
      /** Gets the URL pattern for this route, if it can be statically determined. */
      string getUrlPattern() { result = super.getUrlPattern() }

      /**
       * Gets a function that will handle incoming requests for this route, if any.
       *
       * NOTE: This will be modified in the near future to have a `RequestHandler` result, instead of a `Method`.
       */
      Method getARequestHandler() { result = super.getARequestHandler() }

      /**
       * Gets a parameter that will receive parts of the url when handling incoming
       * requests for this route, if any. These automatically become a `RemoteFlowSource`.
       */
      Parameter getARoutedParameter() { result = super.getARoutedParameter() }

      /** Gets a string that identifies the framework used for this route setup. */
      string getFramework() { result = super.getFramework() }
    }

    /** Provides a class for modeling new HTTP routing APIs. */
    module RouteSetup {
      /**
       * A data-flow node that sets up a route on a server.
       *
       * Extend this class to model new APIs. If you want to refine existing API models,
       * extend `RouteSetup` instead.
       */
      abstract class Range extends DataFlow::Node {
        /** Gets the argument used to set the URL pattern. */
        abstract DataFlow::Node getUrlPatternArg();

        /** Gets the URL pattern for this route, if it can be statically determined. */
        string getUrlPattern() {
          exists(CfgNodes::ExprNodes::StringlikeLiteralCfgNode strNode |
            this.getUrlPatternArg().getALocalSource() = DataFlow::exprNode(strNode) and
            result = strNode.getExpr().getValueText()
          )
        }

        /**
         * Gets a function that will handle incoming requests for this route, if any.
         *
         * NOTE: This will be modified in the near future to have a `RequestHandler` result, instead of a `Method`.
         */
        abstract Method getARequestHandler();

        /**
         * Gets a parameter that will receive parts of the url when handling incoming
         * requests for this route, if any. These automatically become a `RemoteFlowSource`.
         */
        abstract Parameter getARoutedParameter();

        /** Gets a string that identifies the framework used for this route setup. */
        abstract string getFramework();
      }
    }

    /**
     * A function that will handle incoming HTTP requests.
     *
     * Extend this class to refine existing API models. If you want to model new APIs,
     * extend `RequestHandler::Range` instead.
     */
    class RequestHandler extends Method instanceof RequestHandler::Range {
      /**
       * Gets a parameter that could receive parts of the url when handling incoming
       * requests, if any. These automatically become a `RemoteFlowSource`.
       */
      Parameter getARoutedParameter() { result = super.getARoutedParameter() }

      /** Gets a string that identifies the framework used for this route setup. */
      string getFramework() { result = super.getFramework() }
    }

    /** Provides a class for modeling new HTTP request handlers. */
    module RequestHandler {
      /**
       * A function that will handle incoming HTTP requests.
       *
       * Extend this class to model new APIs. If you want to refine existing API models,
       * extend `RequestHandler` instead.
       *
       * Only extend this class if you can't provide a `RouteSetup`, since we handle that case automatically.
       */
      abstract class Range extends Method {
        /**
         * Gets a parameter that could receive parts of the url when handling incoming
         * requests, if any. These automatically become a `RemoteFlowSource`.
         */
        abstract Parameter getARoutedParameter();

        /** Gets a string that identifies the framework used for this request handler. */
        abstract string getFramework();
      }
    }

    private class RequestHandlerFromRouteSetup extends RequestHandler::Range {
      RouteSetup rs;

      RequestHandlerFromRouteSetup() { this = rs.getARequestHandler() }

      override Parameter getARoutedParameter() {
        result = rs.getARoutedParameter() and
        result = this.getAParameter()
      }

      override string getFramework() { result = rs.getFramework() }
    }

    /** A parameter that will receive parts of the url when handling an incoming request. */
    private class RoutedParameter extends RemoteFlowSource::Range, DataFlow::ParameterNode {
      RequestHandler handler;

      RoutedParameter() { this.getParameter() = handler.getARoutedParameter() }

      override string getSourceType() { result = handler.getFramework() + " RoutedParameter" }
    }

    /**
     * A data-flow node that creates a HTTP response on a server.
     *
     * Note: we don't require that this response must be sent to a client (a kind of
     * "if a tree falls in a forest and nobody hears it" situation).
     *
     * Extend this class to refine existing API models. If you want to model new APIs,
     * extend `HttpResponse::Range` instead.
     */
    class HttpResponse extends DataFlow::Node instanceof HttpResponse::Range {
      /** Gets the data-flow node that specifies the body of this HTTP response. */
      DataFlow::Node getBody() { result = super.getBody() }

      /** Gets the mimetype of this HTTP response, if it can be statically determined. */
      string getMimetype() { result = super.getMimetype() }
    }

    /** Provides a class for modeling new HTTP response APIs. */
    module HttpResponse {
      /**
       * A data-flow node that creates a HTTP response on a server.
       *
       * Note: we don't require that this response must be sent to a client (a kind of
       * "if a tree falls in a forest and nobody hears it" situation).
       *
       * Extend this class to model new APIs. If you want to refine existing API models,
       * extend `HttpResponse` instead.
       */
      abstract class Range extends DataFlow::Node {
        /** Gets the data-flow node that specifies the body of this HTTP response. */
        abstract DataFlow::Node getBody();

        /** Gets the data-flow node that specifies the content-type/mimetype of this HTTP response, if any. */
        abstract DataFlow::Node getMimetypeOrContentTypeArg();

        /** Gets the default mimetype that should be used if `getMimetypeOrContentTypeArg` has no results. */
        abstract string getMimetypeDefault();

        /** Gets the mimetype of this HTTP response, if it can be statically determined. */
        string getMimetype() {
          exists(CfgNodes::ExprNodes::StringlikeLiteralCfgNode strNode |
            this.getMimetypeOrContentTypeArg().getALocalSource() = DataFlow::exprNode(strNode) and
            result = strNode.getExpr().getValueText().splitAt(";", 0)
          )
          or
          not exists(this.getMimetypeOrContentTypeArg()) and
          result = this.getMimetypeDefault()
        }
      }
    }

    /**
     * A data-flow node that creates a HTTP redirect response on a server.
     *
     * Note: we don't require that this redirect must be sent to a client (a kind of
     * "if a tree falls in a forest and nobody hears it" situation).
     *
     * Extend this class to refine existing API models. If you want to model new APIs,
     * extend `HttpRedirectResponse::Range` instead.
     */
    class HttpRedirectResponse extends HttpResponse instanceof HttpRedirectResponse::Range {
      /** Gets the data-flow node that specifies the location of this HTTP redirect response. */
      DataFlow::Node getRedirectLocation() { result = super.getRedirectLocation() }
    }

    /** Provides a class for modeling new HTTP redirect response APIs. */
    module HttpRedirectResponse {
      /**
       * A data-flow node that creates a HTTP redirect response on a server.
       *
       * Note: we don't require that this redirect must be sent to a client (a kind of
       * "if a tree falls in a forest and nobody hears it" situation).
       *
       * Extend this class to model new APIs. If you want to refine existing API models,
       * extend `HttpResponse` instead.
       */
      abstract class Range extends HTTP::Server::HttpResponse::Range {
        /** Gets the data-flow node that specifies the location of this HTTP redirect response. */
        abstract DataFlow::Node getRedirectLocation();
      }
    }
  }
}

/**
 * A data flow node that executes an operating system command,
 * for instance by spawning a new process.
 */
class SystemCommandExecution extends DataFlow::Node {
  SystemCommandExecution::Range range;

  SystemCommandExecution() { this = range }

  /** Holds if a shell interprets `arg`. */
  predicate isShellInterpreted(DataFlow::Node arg) { range.isShellInterpreted(arg) }

  /** Gets an argument to this execution that specifies the command or an argument to it. */
  DataFlow::Node getAnArgument() { result = range.getAnArgument() }
}

module SystemCommandExecution {
  /**
   * A data flow node that executes an operating system command, for instance by spawning a new
   * process.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `SystemCommandExecution` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Gets an argument to this execution that specifies the command or an argument to it. */
    abstract DataFlow::Node getAnArgument();

    /** Holds if a shell interprets `arg`. */
    predicate isShellInterpreted(DataFlow::Node arg) { none() }
  }
}
