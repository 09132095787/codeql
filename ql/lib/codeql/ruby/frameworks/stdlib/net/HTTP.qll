private import codeql.ruby.AST
private import codeql.ruby.Concepts
private import codeql.ruby.dataflow.RemoteFlowSources
private import codeql.ruby.ApiGraphs
private import codeql.ruby.dataflow.internal.DataFlowPublic

/**
 * A shortcut for uses of Net::HTTP
 */
private API::Node netHTTP() { result = API::getTopLevelMember("Net").getMember("HTTP") }

/**
 * A call that returns the response body of a `Net::HTTP` request as a String.
 * ```ruby
 * req = Net::HTTP.new("example.com")
 * response = req.get("/")
 * body = response.body
 * ```
 */
private class NetHTTPRequestResponseBody extends CallNode {
  DataFlow::CallNode requestCall;

  NetHTTPRequestResponseBody() {
    exists(string methodName, API::Node requestCallNode |
      requestCall = requestCallNode.getAnImmediateUse()
    |
      // Net::HTTP.get(...)
      methodName = "get" and
      requestCallNode = netHTTP().getReturn(methodName) and
      this = requestCall
      or
      // Net::HTTP.post(...).body
      methodName in ["post", "post_form"] and
      requestCallNode = netHTTP().getReturn(methodName) and
      this = requestCallNode.getAMethodCall(["body", "read_body", "entity"])
      or
      // Net::HTTP.new(..).get(..).body
      methodName in [
          "get", "get2", "request_get", "head", "head2", "request_head", "delete", "put", "patch",
          "post", "post2", "request_post", "request"
        ] and
      requestCallNode = netHTTP().getInstance().getReturn(methodName) and
      this = requestCallNode.getAMethodCall(["body", "read_body", "entity"])
    )
  }

  /**
   * Gets the node representing the method call that initiates the request.
   * This may be different from the node which returns the response body.
   */
  DataFlow::Node getRequestCall() { result = requestCall }

  /**
   * Gets the node representing the URL of the request.
   * Currently unused, but may be useful in future, e.g. to filter out certain requests.
   */
  DataFlow::Node getURLArgument() { result = requestCall.getArgument(0) }
}

/**
 * A `Net::HTTP` call which initiates an HTTP request.
 */
class NetHTTPRequest extends HTTP::Client::Request::Range {
  private NetHTTPRequestResponseBody responseBody;

  NetHTTPRequest() { this = responseBody.getRequestCall().asExpr().getExpr() }

  override DataFlow::Node getResponseBody() { result = responseBody }

  override string getFramework() { result = "Net::HTTP" }
}
