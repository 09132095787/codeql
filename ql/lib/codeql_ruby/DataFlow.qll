/**
 * Provides classes for performing local (intra-procedural) and
 * global (inter-procedural) data flow analyses.
 */
module DataFlow {
  import codeql_ruby.dataflow.internal.DataFlowImpl
}
