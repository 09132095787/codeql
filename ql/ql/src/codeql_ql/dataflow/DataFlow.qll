private import codeql_ql.ast.Ast
private import internal.NodesInternal

/**
 * An expression or variable in a formula, including some additional nodes
 * that are not part of the AST.
 */
class Node extends TNode {
  string toString() { none() } // overridden in subclasses
  Location getLocation() { none() } // overridden in subclasses

  /**
   * Gets the underlying `Expr` or `VarDef` node, if this is an `AstNodeNode`.
   */
  AstNode asAstNode() { astNode(result) = this }

  /**
   * Gets the predicate containing this data-flow node.
   *
   * All data-flow nodes belong in exactly one predicate.
   * TODO: select clauses
   */
  Predicate getEnclosingPredicate() { none() } // overridden in subclasses
}

/**
 * A data-flow node based an `Expr` or `VarDef` AST node.
 */
class AstNodeNode extends Node, MkAstNodeNode {
  private AstNode ast;

  AstNodeNode() { this = MkAstNodeNode(ast) }

  override string toString() {
    result = ast.toString()
  }

  override Location getLocation() {
    result = ast.getLocation()
  }

  /** Gets the AST node. */
  AstNode getAstNode() {
    result = ast
  }

  override Predicate getEnclosingPredicate() {
    result = ast.getEnclosingPredicate()
  }
}

/**
 * Gets the data-flow node correspoinding to the given AST node.
 */
pragma[inline]
Node astNode(AstNode node) {
  result = MkAstNodeNode(node)
}

/**
 * A data-flow node representing a variable within a specific scope.
 */
class ScopedVariableNode extends Node, MkScopedVariable {
  private VarDef var;
  private AstNode scope;

  ScopedVariableNode() { this = MkScopedVariable(var, scope) }

  override string toString() {
    result = "Variable '" + var.getName() + "' scoped to " + scope.getLocation().getStartLine() + ":" + scope.getLocation().getStartColumn()
  }

  override Location getLocation() {
    result = scope.getLocation()
  }

  /** Gets the variable being refined to a specific scope. */
  VarDef getVariable() {
    result = var
  }

  /** Gets the scope to which this variable has been refined. */
  AstNode getScope() {
    result = scope
  }

  override Predicate getEnclosingPredicate() {
    result = var.getEnclosingPredicate()
  }
}

/**
 * Gets the data-flow node corresponding to `var` restricted to `scope`.
 */
pragma[inline]
Node scopedVariable(VarDef var, AstNode scope) {
  result = MkScopedVariable(var, scope)
}

/**
 * A data-flow node representing `this` within a class predicate, charpred, or newtype branch.
 */
class ThisNode extends Node, MkThisNode {
  private Predicate pred;

  ThisNode() { this = MkThisNode(pred) }

  override string toString() {
    result = "'this' in " + pred.getName()
  }

  override Location getLocation() {
    result = pred.getLocation()
  }

  /** Gets the class predicate, charpred, or newtype branch whose 'this' parameter is represented by this node. */
  Predicate getPredicate() {
    result = pred
  }

  override Predicate getEnclosingPredicate() {
    result = pred
  }
}

/**
 * Gets the data-flow node representing `this` within the given class predicate, charpred, or newtype branch.
 */
pragma[inline]
Node thisNode(Predicate pred) {
  result = MkThisNode(pred)
}

/**
 * A data-flow node representing `result` within a predicate that has a result.
 */
class ResultNode extends Node, MkResultNode {
  private Predicate pred;

  ResultNode() { this = MkResultNode(pred) }

  override string toString() {
    result = "'result' in " + pred.getName()
  }

  override Location getLocation() {
    result = pred.getLocation()
  }

  /** Gets the predicate whose 'result' parameter is represented by this node. */
  Predicate getPredicate() {
    result = pred
  }

  override Predicate getEnclosingPredicate() {
    result = pred
  }
}

/**
 * Gets the data-flow node representing `result` within the given predicate.
 */
pragma[inline]
Node resultNode(Predicate pred) {
  result = MkResultNode(pred)
}

/**
 * A data-flow node representing the view of a field in the enclosing class, as seen
 * from a charpred or class predicate.
 */
class FieldNode extends Node, MkFieldNode {
  private Predicate pred;
  private FieldDecl fieldDecl;

  FieldNode() { this = MkFieldNode(pred, fieldDecl) }

  /** Gets the member predicate or charpred for which this node represents access to the field. */
  Predicate getPredicate() {
    result = pred
  }

  FieldDecl getFieldDeclaration() {
    result = fieldDecl
  }

  string getFieldName() {
    result = fieldDecl.getName()
  }

  override string toString() {
    result = "'"  + this.getFieldName() + "' in " + pred.getName()
  }

  override Location getLocation() {
    result = pred.getLocation()
  }

  override Predicate getEnclosingPredicate() {
    result = pred
  }
}

/**
 * Gets the data-flow node representing the given predicate's view of the given field
 * in the enclosing class.
 */
pragma[inline]
Node fieldNode(Predicate pred, FieldDecl fieldDecl) {
  result = MkFieldNode(pred, fieldDecl)
}
