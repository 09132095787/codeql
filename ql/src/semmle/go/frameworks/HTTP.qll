/**
 * Provides classes for working with HTTP-related concepts such as requests and responses.
 */

import go

private module StdlibHttp {
  /** An access to an HTTP request field whose value may be controlled by an untrusted user. */
  private class UserControlledRequestField extends UntrustedFlowSource::Range, DataFlow::FieldReadNode {
    UserControlledRequestField() {
      exists(Type req, string fieldName |
        req.hasQualifiedName("net/http", "Request") and
        this.getField() = req.getField(fieldName) |
        fieldName = "Body" or fieldName = "Form" or fieldName = "Header" or fieldName = "URL"
      )
    }
  }

  private class HeaderGetCall extends UntrustedFlowSource::Range, DataFlow::MethodCallNode {
    HeaderGetCall() {
      this.getTarget().hasQualifiedName("net/http", "Header", "Get")
    }
  }


  private class StdlibResponseWriter extends HTTP::ResponseWriter::Range {
    StdlibResponseWriter() { this.getType().implements("net/http", "ResponseWriter") }

    /** Gets a header object that corresponds to this HTTP response. */
    DataFlow::MethodCallNode getAHeaderObject() {
      result.getTarget().hasQualifiedName("net/http", _, "Header") and
      this.getARead() = result.getReceiver()
    }
  }

  private class HeaderWriteCall extends HTTP::HeaderWrite::Range, DataFlow::MethodCallNode {
    HeaderWriteCall() {
      this.getTarget().hasQualifiedName("net/http", "Header", "Add") or
      this.getTarget().hasQualifiedName("net/http", "Header", "Set")
    }

    override DataFlow::Node getName() { result = this.getArgument(0) }

    override DataFlow::Node getValue() { result = this.getArgument(1) }

    override HTTP::ResponseWriter getResponseWriter() {
      // find `v` in
      // ```
      // header := v.Header()
      // header.Add(...)
      // ```
      result.(StdlibResponseWriter).getAHeaderObject().getASuccessor*() = this.getReceiver()
    }
  }

  private class MapWrite extends HTTP::HeaderWrite::Range, DataFlow::Node {
    Write write;
    DataFlow::Node index;
    DataFlow::Node rhs;

    MapWrite() {
      this.getType().hasQualifiedName("net/http", "Header") and
      write.writesElement(this, index, rhs)
    }

    override DataFlow::Node getName() { result = index }

    override DataFlow::Node getValue() { result = rhs }

    override HTTP::ResponseWriter getResponseWriter() {
      // find `v` in
      // ```
      // header := v.Header()
      // header[...] = ...
      // ```
      result.(StdlibResponseWriter).getAHeaderObject().getASuccessor*() = this
    }
  }

  private class ResponseWriteHeaderCall extends HTTP::HeaderWrite::Range, DataFlow::MethodCallNode {
    ResponseWriteHeaderCall() {
      this.getTarget().implements("net/http", "ResponseWriter", "WriteHeader")
    }

    override string getHeaderName() { result = "status" }

    override predicate definesHeader(string header, string value) {
      header = "status" and value = this.getValue().getIntValue().toString()
    }

    override DataFlow::Node getName() { none() }

    override DataFlow::Node getValue() { result = this.getArgument(0) }

    override HTTP::ResponseWriter getResponseWriter() { result.getANode() = this.getReceiver() }
  }

  private class RequestBody extends HTTP::RequestBody::Range, DataFlow::ExprNode {
    RequestBody() {
      exists(DataFlow::CallNode newRequestCall |
        newRequestCall.getTarget().hasQualifiedName("net/http", "NewRequest")
      |
        this = newRequestCall.getArgument(2)
      )
      or
      exists(Write w, DataFlow::Node base, Field body, Type request |
        w.writesField(base, body, this) and
        request.hasQualifiedName("net/http", "Request") and
        request.getPointerType() = base.getType().getUnderlyingType() and
        body.getName() = "Body"
      )
    }
  }

  private class ResponseBody extends HTTP::ResponseBody::Range, DataFlow::ArgumentNode {
    int arg;

    ResponseBody() {
      exists(DataFlow::CallNode call |
        call.getTarget().(Method).implements("net/http", "ResponseWriter", "Write") and
        arg = 0
        or
        (
          call.getTarget().hasQualifiedName("fmt", "Fprintf")
          or
          call.getTarget().hasQualifiedName("io", "WriteString")
        ) and
        call.getArgument(0).getType().hasQualifiedName("net/http", "ResponseWriter") and
        arg >= 1
      |
        this = call.getArgument(arg)
      )
    }

    override HTTP::ResponseWriter getResponseWriter() {
      // the response writer is the receiver of this call
      result.getANode() = this.getCall().(DataFlow::MethodCallNode).getReceiver()
      or
      // the response writer is an argument to Fprintf or WriteString
      arg >= 1 and
      result.getANode() = this.getCall().getArgument(0)
    }
  }

  private class RedirectCall extends HTTP::Redirect::Range, DataFlow::CallNode {
    RedirectCall() { this.getTarget().hasQualifiedName("net/http", "Redirect") }

    override DataFlow::Node getUrl() { result = this.getArgument(2) }

    override HTTP::ResponseWriter getResponseWriter() { result.getANode() = this.getArgument(0) }
  }
}
