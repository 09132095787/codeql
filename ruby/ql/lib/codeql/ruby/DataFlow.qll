/**
 * Provides classes for performing local (intra-procedural) and
 * global (inter-procedural) data flow analyses.
 */

import codeql.Locations

module DataFlow {
  import codeql.ruby.dataflow.internal.DataFlowImpl
}
