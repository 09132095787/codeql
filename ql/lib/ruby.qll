import codeql_ruby.AST
