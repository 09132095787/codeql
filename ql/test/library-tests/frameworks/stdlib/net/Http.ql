import codeql.ruby.frameworks.stdlib.net.HTTP
import codeql.ruby.DataFlow

query DataFlow::Node netHTTPRequests(NetHTTPRequest e) { result = e.getResponseBody() }
