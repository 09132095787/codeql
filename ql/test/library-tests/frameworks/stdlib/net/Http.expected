| HTTP.rb:4:1:4:18 | call to get | HTTP.rb:4:1:4:18 | call to get |
| HTTP.rb:6:8:6:50 | call to post | HTTP.rb:7:1:7:9 | call to body |
| HTTP.rb:6:8:6:50 | call to post | HTTP.rb:8:1:8:14 | call to read_body |
| HTTP.rb:6:8:6:50 | call to post | HTTP.rb:9:1:9:11 | call to entity |
| HTTP.rb:13:6:13:17 | call to get | HTTP.rb:18:1:18:7 | call to body |
| HTTP.rb:14:6:14:18 | call to post | HTTP.rb:19:1:19:12 | call to read_body |
| HTTP.rb:15:6:15:17 | call to put | HTTP.rb:20:1:20:9 | call to entity |
| HTTP.rb:24:3:24:33 | call to get | HTTP.rb:27:1:27:28 | call to body |
